### ISSUE TITLE
ANR : Input dispatching timed out 

### ISSUE THREAD

**Issue by aditya-keri-betterhalf** [software-mansion/react-native-reanimated#2812]
## Description

**Only** in our Android bundle app (works fine in development mode and in the apk), we are seeing a couple of ANRs/crashes that are occurring when we are scrolling on a particular screen.

The two ANRs are : 

**ANR 1 :** 
```ANR:  Input dispatching timed out (com.****/com.****.MainActivity, Waiting to send non-key event because the touched window has not finished processing certain input events that were delivered to it over 500.0ms ago.  Wait queue length: 8.  Wait queue head age: 9044.1ms.)         
at android.os.MessageQueue.enqueueMessage(MessageQueue.java:584)         
at android.os.Handler.enqueueMessage(Handler.java:754)         
at android.os.Handler.sendMessageAtTime(Handler.java:703)         
at android.os.Handler.sendMessageDelayed(Handler.java:673)        
at android.os.Handler.post(Handler.java:403)         
at com.facebook.react.bridge.queue.MessageQueueThreadImpl.runOnQueue(MessageQueueThreadImpl.java:66)         
at com.facebook.react.bridge.ReactContext.runOnUiQueueThread(ReactContext.java:333)         
at com.swmansion.reanimated.Scheduler.scheduleOnUI(Scheduler.java:37)        
at com.swmansion.reanimated.NativeProxy$AnimationFrameCallback.onAnimationFrame(NativeProxy.java:-2)         
at com.swmansion.reanimated.NodesManager.onAnimationFrame(NodesManager.java:252)         
at com.swmansion.reanimated.NodesManager.access$000(NodesManager.java:65)         
at com.swmansion.reanimated.NodesManager$1.doFrameGuarded(NodesManager.java:153)         
at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:29)         
at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:175)         at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:85)         
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1029)         
at android.view.Choreographer.doCallbacks(Choreographer.java:854)         
at android.view.Choreographer.doFrame(Choreographer.java:785)         
at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1016)         
at android.os.Handler.handleCallback(Handler.java:883)         
at android.os.Handler.dispatchMessage(Handler.java:100)         
at android.os.Looper.loop(Looper.java:224)         
at android.app.ActivityThread.main(ActivityThread.java:7562)         
at java.lang.reflect.Method.invoke(Method.java:-2)         
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)         
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
```

**ANR 2** 
```
ANR:  Input dispatching timed out (1699cb com.****/com.****.MainActivity (server) is not responding. Waited 5001ms for MotionEvent)         
at android.os.Handler.sendMessageAtTime(Handler.java:720)         
at android.os.Handler.sendMessageDelayed(Handler.java:697)         
at android.os.Handler.post(Handler.java:427)         
at com.facebook.react.bridge.queue.MessageQueueThreadImpl.runOnQueue(MessageQueueThreadImpl.java:66)         
at com.facebook.react.bridge.ReactContext.runOnUiQueueThread(ReactContext.java:333)         
at com.swmansion.reanimated.Scheduler.scheduleOnUI(Scheduler.java:37)        
at com.swmansion.reanimated.NativeProxy$AnimationFrameCallback.onAnimationFrame(NativeProxy.java:-2)         
at com.swmansion.reanimated.NodesManager.onAnimationFrame(NodesManager.java:252)         
at com.swmansion.reanimated.NodesManager.access$000(NodesManager.java:65)        
 at com.swmansion.reanimated.NodesManager$1.doFrameGuarded(NodesManager.java:153)         
at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:29)         
at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:175)        
at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:85)         
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1350)         
at android.view.Choreographer.doCallbacks(Choreographer.java:1149)         
at android.view.Choreographer.doFrame(Choreographer.java:1040)         
at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1333)         
at android.os.Handler.handleCallback(Handler.java:938)         
at android.os.Handler.dispatchMessage(Handler.java:99)         
at android.os.Looper.loop(Looper.java:233)        
at android.app.ActivityThread.main(ActivityThread.java:8068)         
at java.lang.reflect.Method.invoke(Method.java:-2)         
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:631)        
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:978)
```

### Expected behavior
App shouldn't crash.

### Actual behavior & steps to reproduce
Whenever we scroll on a particular screen, the app crashes after sometime.
Solutions we've tried so far : 
1. Upgrading reanimated from 2.2.2 to 2.3.1. This just created this issue : https://github.com/software-mansion/react-native-reanimated/issues/2702. Despite trying all the steps provided in the linked issue, the issue still persisted.
2. Upgrading reanimated to 2.2.4. That is when we saw the above ANR.

## Snack or minimal code example

No repro.

## Package versions

| name                                 | version |
| ------------------------------------ | ------- |
| react-native                         | 0.64.2        |
| react-native-reanimated              | 2.2.4        |
| NodeJS                               |    14.17.5     |
| Xcode                                |   12.5.1      |

## Affected platforms

- [x] Android
- [ ] iOS
- [ ] Web

**Comment by wayne-tt** [comment]
I got a similar error. I'm guessing that There's a bug within the CPP code. scheduleOnUI is called, but it seems that an invalid pointer value is passed. and I'm guessing it's occupying the UI thread for a long time.

**Comment by kmagiera** [comment]
Thanks @aditya-keri-betterhalf for reporting. Can you make sure the posted traces contain full details about the ANRs? It seem like they might not be complete, in particular the first one. They may also contain traces for the native code (c++)? If so please include these.

On top of that since it seems like you can easily reproduce the issue, it would be extremely helpful if you could try to come up with a minimum repro scenario in a form of a github repo that we could try to reproduce the problem. The best way to work on such an example is by trying to disable/comment out fragments of you app so long the issue persist.

**Comment by kmagiera** [comment]
Also thanks @wayne-tt for your comments, can you elaborate what makes you think this is related to invalid pointer passed to scheduleOnUI?

**Comment by henrikra** [comment]
@kmagiera please check this similar ticket https://github.com/software-mansion/react-native-reanimated/issues/2832

**Comment by wayne-tt** [comment]
My issue was caused by excessive use of useAnimatedReaction. After I replaced useAnimatedReaction with another hook, the ANR was gone.

**Comment by henrikra** [comment]
> My issue was caused by excessive use of useAnimatedReaction. After I replaced useAnimatedReaction with another hook, the ANR was gone.

We are getting lot of this same ANR and we don't have any `useAnimatedReaction` in the code

**Comment by sregg** [comment]
We're seeing similar ANRs on both Android and iOS using this library: https://github.com/Flair-Dev/react-native-awesome-gallery which uses `useAnimatedReaction`.
We're using reanimated v2.9.1 but it also happens with v2.10.0.

**Comment by efstathiosntonas** [comment]
In my case the only library that uses `useAnimatedReaction` is [react-native-awesome-slider](https://www.npmjs.com/package/react-native-awesome-slider). I get about 500 ANR per month...

**Comment by Latropos** [comment]
Closing since fix is merged #4239
